src/Sylius/Component/Taxation/Checker/TaxRateDateChecker.php

```diff
@@ -0,0 +1,51 @@
+<?php
+
+/*
+ * This file is part of the Sylius package.
+ *
+ * (c) Paweł Jędrzejewski
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+declare(strict_types=1);
+
+namespace Sylius\Component\Taxation\Checker;
+
+use DateTimeInterface;
+use Sylius\Calendar\Provider\DateTimeProviderInterface;
+use Sylius\Component\Taxation\Model\TaxRateInterface;
+
+final class TaxRateDateChecker implements TaxRateDateCheckerInterface
+{
+    public function __construct(
+        protected DateTimeProviderInterface $calendar,
+    ) {
+    }
+
+    public function filter(array $taxRates): ?TaxRateInterface
+    {
+        $taxRates = array_filter($taxRates, function ($taxRate){
+            if ($this->isInDate($this->calendar->now(), $taxRate)) {
+                return $taxRate;
+            }
+        });
+
+        $taxRates = array_values($taxRates);
+
+        return $taxRates[0] ?? null;
```

Comment:
Just replacing the foreach with an array_filter is not an improvement, especially when doing it wrong (the callable should return true/false, not objects).
An improvement would be to bring back the foreach, but with an early return.
That's because there's no point to iterate over all tax rates, you only need the first one that's eligible.
```suggestion
        foreach ($taxRates as $taxRate) {
            if ($this->isInDate($this->calendar->now(), $taxRate)) {
                return $taxRate;
            }
        }

        return null;
```